### ISSUE TITLE
Combined Comment & subscribe flow

### ISSUE THREAD

**Issue by SidharthBansal** [publiclab/plots2#3932]
Combined “Comment + subscribe” flow - get people subscribed to some tags as they leave a comment, maybe based on the tags of the post they're commenting on, perhaps using parameters to the comment post action
Refer #3924

**Comment by SidharthBansal** [comment]
What I am thinking is a little complex. Please try to think in that fashion and if it is not clear please ask me again. 
Suppose we have successfully created ``Comment+Subscribe`` flow. The system is working properly. 
Whenever a person leaves a comment, they are automatically subscribed to some of the tags on some criterion we have selected in our implementation. Public Lab's folks are all happy. 

There will be many existing notes, comments and tags in the system which are present. But are not in accordance with the functionality. 
I mean there will be a user U1's research note R1 with tag T1 and comment C1 by the user U2 before the implementation of the #3932. So, do we need to subscribe tag T1 to the user U2?
If we do subscriptions for the previous database state also it will have the following two drawbacks
1) Users will be highly dissatisfied with the abrupt increase in the # of emails they are getting from public lab
2) We need to do subscriptions for the entire previous state of database. This will be a time consuming job.
@jywarren do we need to see this?

**Comment by jywarren** [comment]
Yes, i think we should start with this going forward only. I'm also
cautious about email noise and people being unhappy.

One thing we might do is to have a way to email anyone who's posted on a
topic /just once/ to see if we can get them to subscribe to that topic. But
this is a big project and I think separate from this issue. We might even
(to avoid sending lots of "please subscribe" emails) think of a way we can
highlight topics you've posted on a lot on your dashboard and say like
"You've posted 5 times on air-quality -- click here to subscribe!"

**Comment by jywarren** [comment]
I think we could start with something like posting a comment, but when you click Post, it says "You need to log in first." and offers a button to log in via OAuth. To the user, this will be like logging in, even though really it is signing up.

Either this happens in a modal, and when it disappears, they are back to their comment box with the comment to submit, OR, we store the contents of the comment box and submit it upon completing the OAuth log in. 

What do you think of that, it might be fewer possible routes if we do it one of these two ways.

**Comment by SidharthBansal** [comment]
I was also thinking of the same behaviour. I completely agree with the https://github.com/publiclab/plots2/issues/3932#issuecomment-438875615. 

One important thing is we have many tags on the research notes and users have existing subscriptions of tags. So, some tags which the `comments+subscribe flow` will subscribe to the user will be conflicting with the existing user tags. So, we need to create a filter to take out only those tags which are not subscribed to the user. Does this make sense?
If you agree I will create an issue and solve it there.
Let us say the tag list of the research note is `Note_tag_list` and user `U` has tag_list `User_tag_list`. 
For better understanding, 
Psuedocode: Let's name the filter as subscribable_tags
Subscribable_tags(Note_tag_list, User_tag_list, U)
   result = {}
   for each tag T in Note_tag_list
       if T is not present in User_tag_list
           result = result + T
    return T


We have ``two possible solutions`` for #3932 : 
1)``Manual comment+subscribe flow``: Show the users the tags which he/she can subscribe to on the comment after applying the filter discussed above.
Disadvantage: Website becomes too complex for user experience. As a starting user, you see multiple tags and get perplexed which tag to subscribe and which not to.
Also, tag duplication. We have tags on the left hand side of each note. If we will also list the tags in the comment section after filtering, this will result in ambiguity to the user. User can't find out why the tags are present on the left hand side but not on the comments. (User doesn't know about the filter method. Users may not be programmers)
In case we do not use filter and use a message to notify user that this tag is already subscribed under each comment to be posted, this will result in redundancy.
So, ``I am not in favour`` of this approach. What do you say @jywarren?

2) ``Automatic comment+subscribe flow``: Automatically subscribe a person whenever they write a comment. Obviously, we want a filter method for this as discussed above.
Let's discuss this. Let us say that we designed some method M which tells us which tags #3932 subscribe to upon posting a comment. Model M is designed to give top 5 comments which we can get subscribed to the user who posted the comment. We have two sub-approaches in (2) as listed below:
2a)Filter then use M: Suppose there is a user like @jywarren who uses public lab on daily basis. We are feeding the subscribable_tags into the model each time that the user posts a comment. This will eventually result in all the tags being subscribed to the user. So, email flood. (I know you are always overwhelmed by the emails you get. But you are superb in replying back to all :1st_place_medal: :100:  )
Email flood may cause the user to discontinue PL.
2b)Use M then filter: M will give some tags which have high affinity towards that article/user. We will send these tags to the filter subscribable_tags. This will result in some tag list(may be of zero length if all the tags suggested by M are taken up by the user). We can subscribe to them.
This will never result in all tags being subscribed to the user, if user does frequent activities on PL.
``I am in favour of 2b approaches.``
What do you say?

**Comment by SidharthBansal** [comment]
@jywarren any views?

**Comment by SidharthBansal** [comment]
Also, M can be trending tags. But I think we need to optimise it. What do you say?

Or do we need to create a separate function for this ?

**Comment by jywarren** [comment]
OK sorry going through these now!

> So, some tags which the comments+subscribe flow will subscribe to the user will be conflicting with the existing user tags. 

We can deduplicate as we do in this code: https://github.com/publiclab/plots2/blob/46e2447dab9948f08e78ca074c2df382acdcb9fd/app/controllers/tag_controller.rb#L291-L296 👍 

😛 LOL `Suppose there is a user like @jywarren who uses public lab on daily basis.`

To your bigger question, I think we can use the idea of "strong user interface defaults" to encourage subscription but still make it optional. For example:

![screen shot 2018-11-30 at 5 21 15 pm](https://user-images.githubusercontent.com/24359/49317906-603a4900-f4c4-11e8-8bd5-52e9731a71fc.png)


(see mockup here: https://docs.google.com/presentation/d/1TCZoTfuhamRVrUak8aDgqJAwSgyhRtZg2Pgacl2_4zc/edit#slide=id.g4877128470_0_0)

**Comment by SidharthBansal** [comment]
I liked your suggestion for `` tags from the comment's parent node which have the most followers`` than my suggestion of trending tags.
Trending tags are always visible to the user at many web pages. So, the user will be able to subscribe them from those web pages, if they wish.

How many tags should a user be able to subscribe on a comment? I prefer at most 1. We don't want email floods from Public Lab in their email accounts.

**Comment by SidharthBansal** [comment]
@jywarren even though I have completed this issue half in the PR. But I have a suggestion. This suggestion might result in closing of this issue.
![image](https://user-images.githubusercontent.com/20972099/50247860-1a0b4200-03ff-11e9-8598-c5cb295e0e0f.png)
![image](https://user-images.githubusercontent.com/20972099/50247878-242d4080-03ff-11e9-89ff-bf2841825dbd.png)

When you compare the two images and their sidebars. You will feel that we are having white portions on the right. It would be much better if the sidebar is static and the main content can be scrolled up and down.
If we implement this design then, there is no need to show tag subscriptions in the comments section as this will introduce redundancy.
What are your views?

**Comment by jywarren** [comment]
Hmm, interesting! Are there any situations where sidebar content would be
taller than one screen?

**Comment by SidharthBansal** [comment]
Yes there are.
I found them but I missed to tell. There are some posts in which number of tags are really large. Also, it depends on the zooming the screen. Some prefer smaller size, some prefer larger size. I think now we should let the sidebar be as it is.
Thanks Jeff.

**Comment by SidharthBansal** [comment]
Cool...

**Comment by SidharthBansal** [comment]
@jywarren I think we should not plan for this issue due to the following reasons:
* High redundancy on the page as each time a person writes a comment, they get subscription prompts. It is annoying to have the same action in case a user has already done it on a particular node(note/ map/ question/ wiki).
* User is getting two buttons on each node. One in the sidebar already implemented and also on the comments section. This creates ambiguity for the user. Are these things different? 
* User want simple UI. Especially the first timers on the Public Lab website. They need to study what tags and power tags are before submitting a comment. They will think what are these checkboxes for. So, it will be a barrier to the users.
* Redundancy on the page as the subscription prompt is on both RHS(comment section) and LHS(Sidebar). Redundancy will be more for initial comments when both are viewed simultaneously.

I don't think, this is a preferable place to add subscription prompt. Sorry this is very late stage when I came across this suggestion. But I don't want we will do something which will be barrier to upcoming or existing users at PL.
What are your views @jywarren @ebarry @steviepubliclab ?

**Comment by jywarren** [comment]
I think this is reasonable. What if we swapped it for a project to identify
tags people have been active on, but aren't subscribed to -- something like
`user.active_topics(limit)` -- which could return a max of `limit` tags for:

1. tags user is not subscribed to
2. which occur a lot on nodes they've posted
3. which occur a lot on nodes on which they've commented

Then we'd have a list of likely candidates to prompt them to subscribe to.
This is an unrelated project but could be swapped in for this one (since I
agree with your comment above).

Then we could use `user.active_topics` to prompt users in places like the
dashboard or tags page, with something like "You've been posting a lot on
[air-quality]. Click here to follow that topic."



This could be implemented similarly to Tag.related() code in the tag.rb
model.

**Comment by jywarren** [comment]
OK i'm going to copy that over to #4644 as it would be usable on the dashboard! We can close this if you want!

**Comment by SidharthBansal** [comment]
Hi Jeff, I liked your suggestion. But can you please tell what you mean by swapping??
Comment + Subscribe and active_topics is entirely different things.
 I think we should open up a new issue for it instead of #4644. 
Do you mean that the Topics section's tags to be replaced by this function?
Also what is the current logic of displaying tags on dashboard?

**Comment by jywarren** [comment]
Yes, sorry - by swapping i mean like, let's do that instead of this!


Currently they're manually curated to represent important environmental
issues. So I think we should keep the Topics section tags as they are for
now, and instead add an additional box below with this new idea. Thank you!
